File: python_src/cartridge_sdk/ui/detail_view.py
"""Scrollable word-wrapped text view."""

from __future__ import annotations

from typing import Callable, TYPE_CHECKING

import pygame

from cartridge_sdk.input import Button, InputEvent
from cartridge_sdk.screen import _get_font
from cartridge_sdk.ui.widget import Widget

if TYPE_CHECKING:
    from cartridge_sdk.theme import Theme

PAD_X = 12
PAD_Y = 8


class DetailView(Widget):
    """Scrollable text content with word wrap. B to go back."""

    def __init__(
        self,
        title: str = "",
        body: str = "",
        on_back: Callable | None = None,
    ) -> None:
        self.title = title
        self.body = body
        self.on_back = on_back
        self._scroll: int = 0
        self._wrapped_lines: list[tuple[str, str]] = []  # (text, style)
        self._needs_wrap: bool = True
        self._last_width: int = 0

    def set_content(self, title: str, body: str) -> None:
        self.title = title
        self.body = body
        self._scroll = 0
        self._needs_wrap = True

    def handle_input(self, event: InputEvent) -> bool:
        if event.action not in ("press", "repeat"):
            return False

        if event.button == Button.B:
            if self.on_back:
                self.on_back()
            return True
        if event.button == Button.DPAD_UP:
            self._scroll = max(0, self._scroll - 1)
            return True
        if event.button == Button.DPAD_DOWN:
            self._scroll += 1
            return True
        if event.button == Button.L2:
            self._scroll = max(0, self._scroll - 10)
            return True
        if event.button == Button.R2:
            self._scroll += 10
            return True

        return False

    def draw(self, surface: pygame.Surface, rect: pygame.Rect, theme: Theme) -> None:
        pygame.draw.rect(surface, theme.bg, rect)

        content_width = rect.width - PAD_X * 2 - 8  # scrollbar margin

        # Re-wrap if needed
        if self._needs_wrap or content_width != self._last_width:
            self._wrapped_lines = self._wrap_text(content_width, theme)
            self._last_width = content_width
            self._needs_wrap = False

        font = _get_font("mono", 14)
        font_bold = _get_font("mono_bold", 16)
        lh = font.get_linesize()
        lh_bold = font_bold.get_linesize()

        # Visible lines
        visible_lines = max(1, (rect.height - PAD_Y * 2) // lh)
        total_lines = len(self._wrapped_lines)
        max_scroll = max(0, total_lines - visible_lines)
        self._scroll = min(self._scroll, max_scroll)

        clip = surface.get_clip()
        surface.set_clip(rect)

        y = rect.y + PAD_Y
        for i in range(self._scroll, min(self._scroll + visible_lines, total_lines)):
            text, style = self._wrapped_lines[i]
            if style == "title":
                rendered = font_bold.render(text, True, theme.text)
                surface.blit(rendered, (rect.x + PAD_X, y))
                y += lh_bold
            elif style == "dim":
                rendered = font.render(text, True, theme.text_dim)
                surface.blit(rendered, (rect.x + PAD_X, y))
                y += lh
            else:
                rendered = font.render(text, True, theme.text)
                surface.blit(rendered, (rect.x + PAD_X, y))
                y += lh

        surface.set_clip(clip)

        # Scroll indicator
        if total_lines > visible_lines:
            ind_x = rect.right - 5
            bar_top = rect.y + PAD_Y
            bar_h = rect.height - PAD_Y * 2
            track_color = (
                min(theme.border[0] + 10, 255),
                min(theme.border[1] + 10, 255),
                min(theme.border[2] + 10, 255),
            )
            pygame.draw.line(surface, track_color, (ind_x, bar_top), (ind_x, bar_top + bar_h))
            thumb_h = max(8, bar_h * visible_lines // total_lines)
            progress = self._scroll / max_scroll if max_scroll > 0 else 0
            thumb_y = bar_top + int((bar_h - thumb_h) * progress)
            pygame.draw.rect(surface, theme.text_dim, (ind_x - 1, thumb_y, 3, thumb_h))

    def _wrap_text(self, max_width: int, theme: Theme) -> list[tuple[str, str]]:
        lines: list[tuple[str, str]] = []
        font = _get_font("mono", 14)
        font_bold = _get_font("mono_bold", 16)

        # Title lines
        if self.title:
            for tline in _word_wrap(self.title, font_bold, max_width):
                lines.append((tline, "title"))
            lines.append(("", "normal"))  # spacer

        # Body lines
        for paragraph in self.body.split("\n"):
            if not paragraph.strip():
                lines.append(("", "normal"))
                continue
            for wline in _word_wrap(paragraph, font, max_width):
                lines.append((wline, "normal"))

        return lines


def _word_wrap(text: str, font: pygame.font.Font, max_width: int) -> list[str]:
    if not text:
        return [""]
    words = text.split(" ")
    lines = []
    current = ""
    for word in words:
        test = f"{current} {word}".strip()
        if font.size(test)[0] <= max_width:
            current = test
        else:
            if current:
                lines.append(current)
            # Handle long words that don't fit
            if font.size(word)[0] > max_width:
                while word:
                    chunk = word
                    while font.size(chunk)[0] > max_width and len(chunk) > 1:
                        chunk = chunk[:-1]
                    lines.append(chunk)
                    word = word[len(chunk):]
                current = ""
            else:
                current = word
    if current:
        lines.append(current)
    return lines or [""]

File: crates/cartridge-core/src/ui/detail_view.rs
//! Scrollable word-wrapped text view with D-pad scrolling.

use sdl2::pixels::Color;
use sdl2::rect::Rect;

use crate::input::{Button, InputAction, InputEvent};
use crate::screen::Screen;

const PAD_X: i32 = 12;
const PAD_Y: i32 = 8;

/// Style for a wrapped line.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum LineStyle {
    Title,
    Normal,
    Dim,
}

/// A single wrapped line with its style.
struct WrappedLine {
    text: String,
    style: LineStyle,
}

/// Scrollable text content with word wrap. D-pad Up/Down scrolls line by line,
/// L1/R1 pages.
pub struct DetailView {
    pub title: String,
    pub body: String,
    scroll: i32,
    wrapped_lines: Vec<WrappedLine>,
    needs_wrap: bool,
    last_width: i32,
}

impl DetailView {
    pub fn new(title: &str, body: &str) -> Self {
        Self {
            title: title.to_string(),
            body: body.to_string(),
            scroll: 0,
            wrapped_lines: Vec::new(),
            needs_wrap: true,
            last_width: 0,
        }
    }

    /// Replace content and reset scroll position.
    pub fn set_content(&mut self, title: &str, body: &str) {
        self.title = title.to_string();
        self.body = body.to_string();
        self.scroll = 0;
        self.needs_wrap = true;
    }

    /// Handle D-pad scroll input. Returns `true` if consumed.
    pub fn handle_input(&mut self, event: &InputEvent) -> bool {
        if event.action == InputAction::Release {
            return false;
        }
        match event.button {
            Button::DpadUp => {
                self.scroll = (self.scroll - 1).max(0);
                true
            }
            Button::DpadDown => {
                self.scroll += 1;
                true
            }
            Button::L1 => {
                self.scroll = (self.scroll - 10).max(0);
                true
            }
            Button::R1 => {
                self.scroll += 10;
                true
            }
            _ => false,
        }
    }

    /// Draw the detail view within `rect`.
    pub fn draw(&mut self, screen: &mut Screen, rect: Rect) {
        let theme = screen.theme;

        screen.draw_rect(rect, Some(theme.bg), true, 0, None);

        let content_width = rect.width() as i32 - PAD_X * 2 - 8; // scrollbar margin

        // Re-wrap if needed
        if self.needs_wrap || content_width != self.last_width {
            self.wrap_text(screen, content_width);
            self.last_width = content_width;
            self.needs_wrap = false;
        }

        let lh = screen.get_line_height(14, false) as i32;
        let lh_bold = screen.get_line_height(16, true) as i32;

        // Visible lines
        let visible_lines = ((rect.height() as i32 - PAD_Y * 2) / lh).max(1);
        let total_lines = self.wrapped_lines.len() as i32;
        let max_scroll = (total_lines - visible_lines).max(0);
        self.scroll = self.scroll.min(max_scroll);

        let mut y = rect.y() + PAD_Y;
        let end = (self.scroll + visible_lines).min(total_lines);
        for i in self.scroll..end {
            let line = &self.wrapped_lines[i as usize];
            match line.style {
                LineStyle::Title => {
                    screen.draw_text(
                        &line.text,
                        rect.x() + PAD_X,
                        y,
                        Some(theme.text),
                        16,
                        true,
                        None,
                    );
                    y += lh_bold;
                }
                LineStyle::Dim => {
                    screen.draw_text(
                        &line.text,
                        rect.x() + PAD_X,
                        y,
                        Some(theme.text_dim),
                        14,
                        false,
                        None,
                    );
                    y += lh;
                }
                LineStyle::Normal => {
                    screen.draw_text(
                        &line.text,
                        rect.x() + PAD_X,
                        y,
                        Some(theme.text),
                        14,
                        false,
                        None,
                    );
                    y += lh;
                }
            }
        }

        // Scroll indicator
        if total_lines > visible_lines {
            let ind_x = rect.x() + rect.width() as i32 - 5;
            let bar_top = rect.y() + PAD_Y;
            let bar_h = rect.height() as i32 - PAD_Y * 2;
            let track_color = Color::RGB(
                theme.border.r.saturating_add(10),
                theme.border.g.saturating_add(10),
                theme.border.b.saturating_add(10),
            );
            screen.draw_line(
                (ind_x, bar_top),
                (ind_x, bar_top + bar_h),
                Some(track_color),
                1,
            );
            let thumb_h = ((bar_h * visible_lines) / total_lines).max(8);
            let progress = if max_scroll > 0 {
                self.scroll as f32 / max_scroll as f32
            } else {
                0.0
            };
            let thumb_y = bar_top + ((bar_h - thumb_h) as f32 * progress) as i32;
            screen.draw_rect(
                Rect::new(ind_x - 1, thumb_y, 3, thumb_h as u32),
                Some(theme.text_dim),
                true,
                0,
                None,
            );
        }
    }

    fn wrap_text(&mut self, screen: &mut Screen, max_width: i32) {
        self.wrapped_lines.clear();

        // Title lines
        if !self.title.is_empty() {
            let title_lines = word_wrap(&self.title, screen, 16, true, max_width);
            for tl in title_lines {
                self.wrapped_lines.push(WrappedLine {
                    text: tl,
                    style: LineStyle::Title,
                });
            }
            // Spacer
            self.wrapped_lines.push(WrappedLine {
                text: String::new(),
                style: LineStyle::Normal,
            });
        }

        // Body lines -- split on newlines for paragraphs.
        // Lines prefixed with "~" are rendered in dim style (secondary text).
        for paragraph in self.body.split('\n') {
            if paragraph.trim().is_empty() {
                self.wrapped_lines.push(WrappedLine {
                    text: String::new(),
                    style: LineStyle::Normal,
                });
                continue;
            }
            // Check for dim prefix
            let (text, style) = if paragraph.starts_with('~') && paragraph.ends_with('~') && paragraph.len() > 2 {
                (&paragraph[1..paragraph.len() - 1], LineStyle::Dim)
            } else {
                (paragraph, LineStyle::Normal)
            };
            let lines = word_wrap(text, screen, 14, false, max_width);
            for line in lines {
                self.wrapped_lines.push(WrappedLine {
                    text: line,
                    style,
                });
            }
        }
    }
}

/// Word-wrap text to fit within `max_width` pixels.
fn word_wrap(
    text: &str,
    screen: &mut Screen,
    font_size: u16,
    bold: bool,
    max_width: i32,
) -> Vec<String> {
    if text.is_empty() {
        return vec![String::new()];
    }

    let words: Vec<&str> = text.split(' ').collect();
    let mut lines = Vec::new();
    let mut current = String::new();

    for word in &words {
        let test = if current.is_empty() {
            word.to_string()
        } else {
            format!("{current} {word}")
        };
        let tw = screen.get_text_width(&test, font_size, bold) as i32;
        if tw <= max_width {
            current = test;
        } else {
            if !current.is_empty() {
                lines.push(current);
            }
            // Handle long words that don't fit
            let word_w = screen.get_text_width(word, font_size, bold) as i32;
            if word_w > max_width {
                let mut remaining = word.to_string();
                while !remaining.is_empty() {
                    let mut chunk = remaining.clone();
                    while screen.get_text_width(&chunk, font_size, bold) as i32 > max_width
                        && chunk.len() > 1
                    {
                        chunk.pop();
                    }
                    let chunk_len = chunk.len();
                    lines.push(chunk);
                    remaining = remaining[chunk_len..].to_string();
                }
                current = String::new();
            } else {
                current = word.to_string();
            }
        }
    }
    if !current.is_empty() {
        lines.push(current);
    }

    if lines.is_empty() {
        vec![String::new()]
    } else {
        lines
    }
}
